students/adam_strong/lesson09/mailroom_oo/donor_models.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+
+""" Donor classes to be run by cli_main.py
+"""
+
+# Starting data structure of donors:
+
+import copy
+import pathlib
+
+donor_db = {"Scrooge McDuck": [8000.00, 70000.00],
+            "Montgomery Burns": [49.53],
+            "Richie Rich": [1000000.00, 500000.00],
+            "Chet Worthington": [200.00, 44387.63, 10200.00],
+            "Silas Skinflint": [0.25, 1.00, 0.43]}
+
+report_head = '{:20}| {:>15}|{:>15}| {:>15}'.format(
+    'Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+
+letter = "\n".join(("", "Dear {},", "",
+                        "    Thank you for your very kind donations totaling ${:.2f}.", "",
+                        "    It will be put to very good use.", "",
+                        "               Sincerely,",
+                        "                  - The team"))
+
+break_line = '-'*72
+
+ty_message = "\n".join(("", "Dear {}",
+                            "Thank you for your generous donation of {:.2f}", ""))
+
+repform = '{:20} ${:>15.2f}  {:>15}  ${:15.2f}'
+
+
+class DonorCollections(object):
+    '''DonorCollections class used to store and manipultate the main dict'''
+
+    donors = copy.deepcopy(donor_db)
+
+    def __init__(self, donors=donors):
+        self.donors = donors
+
+    def add_donor(self, tyname, amt):
+        '''Adds a new donor or updates an existing donor, returns thank you letter'''
+        tyname = tyname.title()
+        if self.donors.get(tyname) is None:
+            self.donors[tyname] = [amt]
+        else:
+            self.donors.get(tyname).append(amt)
+        ty_output = ty_message.format(tyname, amt)
+        return ty_output
+
+    def make_report(self):
+        '''Creates and returns the report on the data'''
+        aggregate = ['\n', report_head, break_line]
+        for key, value in sorted(self.donors.items(), key=lambda i: sum(i[1]), reverse=True):
+            don = Donor(key, value)
+            line = don.format_donor()
+            aggregate.append(line)
+        return aggregate
+
+    def send_letter(self):
+        '''Executes the loop to the Donor class to send letters to each donor'''
+        for key, value in sorted(self.donors.items(), key=lambda i: sum(i[1]), reverse=True):
+            don = Donor(key, value)
+            letter = don.create_letter()
```

Comment:
all of these functions would change to reference existing donor instances instead of creating new ones.